src/com/palmergames/bukkit/towny/listeners/TownyPlayerListener.java

```diff
@@ -1003,6 +1003,77 @@ public void onJailedPlayerUsesCommand(PlayerCommandPreprocessEvent event) {
 		}
 	}
 	
+	/** 
+	 * Allows restricting commands while being in a town.
+	 * Also allows limiting commands to personally-owned plots only.
+	 * Works almost the same way as jail command blacklisting, except has more stuff
+	 * @param event PlayerCommandPreprocessEvent
+	 */
+	@EventHandler(priority = EventPriority.NORMAL)
+	public void onPlayerUsesCommandInsideTown(PlayerCommandPreprocessEvent event) throws NotRegisteredException {
+		if (plugin.isError() || !TownySettings.allowTownCommandBlacklisting() ||
+				!TownyAPI.getInstance().isTownyWorld(event.getPlayer().getWorld()))
+			return;
+		
+		Player player = event.getPlayer();
+		Resident res = TownyUniverse.getInstance().getResident(player.getUniqueId());
+		
+		if (res == null || TownyUniverse.getInstance().getPermissionSource().has(player, PermissionNodes.TOWNY_ADMIN_TOWN_COMMAND_BLACKLIST_BYPASS.getNode()))
+			return;
+		
+		String[] split = event.getMessage().substring(1).split(" ");
+		
+		if (TownySettings.getPlayerOwnedPlotLimitedCommands().contains(split[0])) {
```

Comment:
The permission check can be moved to after the two .contains(split[0]) checks, so that it doesn't check for permissions each time a player does any command.